bgs/bgs.go

```diff
@@ -394,26 +413,86 @@ func (bgs *BGS) EventsHandler(c echo.Context) error {
 		since = &sval
 	}
 
-	ctx := c.Request().Context()
+	ctx, cancel := context.WithCancel(c.Request().Context())
+	defer cancel()
 
 	// TODO: authhhh
 	conn, err := websocket.Upgrade(c.Response(), c.Request(), c.Response().Header(), 1<<10, 1<<10)
 	if err != nil {
 		return fmt.Errorf("upgrading websocket: %w", err)
 	}
 
-	evts, cancel, err := bgs.events.Subscribe(ctx, func(evt *events.XRPCStreamEvent) bool { return true }, since)
+	// Start a goroutine to ping the client every 30 seconds to check if it's
+	// still alive. If the client doesn't respond to a ping within 5 seconds,
+	// we'll close the connection and teardown the consumer.
+	go func() {
+		ticker := time.NewTicker(30 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if err := conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(5*time.Second)); err != nil {
+					log.Errorf("failed to ping client: %s", err)
+					cancel()
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	evts, cleanup, err := bgs.events.Subscribe(ctx, func(evt *events.XRPCStreamEvent) bool { return true }, since)
 	if err != nil {
 		return err
 	}
-	defer cancel()
+	defer cleanup()
+
+	// Keep track of the consumer for metrics and admin endpoints
+	consumer := SocketConsumer{
+		RemoteAddr:  c.RealIP(),
+		UserAgent:   c.Request().UserAgent(),
+		ConnectedAt: time.Now(),
+	}
+
+	sentCounter := eventsSentCounter.WithLabelValues(consumer.RemoteAddr, consumer.UserAgent)
```

Comment:
the cardinality on this metric is probably going to be very high

Reply:
I'm guessing it'll be under 10k series initially, basically one series per unique consumer (though repeat consumers will end up using the same series). If it starts to get too high we can try reducing it but I like the idea of remote address and/or user-agent tracking in a visible way.

Reply:
Also it'll get GC'd on BGS restart so we don't have a ton of inactive series to track (hopefully).

Reply:
We could maybe track it internally at this level and via prometheus only at the RemoteAddr level to start

Reply:
yeah i guess that seems fine for now, lets just keep an eye on it